experimental/list_formatter/src/deduplicating_array.rs

```diff
@@ -0,0 +1,226 @@
+use serde::de::{Deserialize, Deserializer, EnumAccess, Visitor};
+use serde::ser::{Serialize, Serializer};
+
+pub fn serialize<S, T, const N: usize>(array: &[T; N], serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+    T: Serialize + PartialEq + Clone,
+{
+    use serde::ser::SerializeTuple;
+
+    let json = serializer.is_human_readable();
+
+    let mut seq = serializer.serialize_tuple(N)?;
+
+    for i in 0..N {
+        match array.iter().take(i).position(|j| &array[i] == j) {
+            None if json => seq.serialize_element(&JsonDedupeSer::Value(&array[i]))?,
+            None => seq.serialize_element(&BincodeDedupeSer::Value(&array[i]))?,
+            Some(j) if json => seq.serialize_element(&JsonDedupeSer::<T>::Fallback([j]))?,
+            Some(j) => seq.serialize_element(&BincodeDedupeSer::<T>::Fallback(j))?,
+        }
+    }
+    seq.end()
+}
+
+pub fn deserialize<'de, D, T, const N: usize>(deserializer: D) -> Result<[T; N], D::Error>
+where
+    D: Deserializer<'de>,
+    T: Deserialize<'de> + Sized + Clone,
+    [JsonDedupeDe<T>; N]: Deserialize<'de>,
+    [BincodeDedupeDe<T>; N]: Deserialize<'de>,
+{
+    use core::mem::MaybeUninit;
+
+    let mut array: [MaybeUninit<T>; N] = unsafe { MaybeUninit::uninit().assume_init() };
+
+    if deserializer.is_human_readable() {
+        for (i, r) in
+            IntoIterator::into_iter(<[JsonDedupeDe<T>; N]>::deserialize(deserializer)?).enumerate()
+        {
+            match r {
+                JsonDedupeDe::Value(v) => {
+                    array[i].write(v);
+                }
+                JsonDedupeDe::Fallback([j]) => unsafe {
+                    // Fallbacks should always be a previous value, which makes the
+                    // assume_init_ref safe
+                    assert!(j < i);
```

Comment:
Nit: Return an error instead of panicking, because we don't want malicious data to be able to crash an app.